crates/engine/src/host_component.rs

```diff
@@ -0,0 +1,89 @@
+use std::{any::Any, marker::PhantomData};
+
+use spin_manifest::CoreComponent;
+use wasmtime::Linker;
+
+use crate::RuntimeContext;
+
+/// Represents a host implementation of a Wasm interface.
+pub trait HostComponent: Send + Sync {
+    /// Host component runtime state.
+    type Data: Any + Send;
+
+    /// Add this component to the given Linker, using the given runtime state-getting closure.
```

Comment:
Not sure I understand the reference to a closure - does this refer to a previous implementation?

Reply:
Woops! Yes indeed; I originally tried to pass a closure in rather than this "handle". I still don't understand why that wasn't essentially equivalent to this, but after much lifetime wrangling I gave up on it.

Reply:
Interesting, I would have expected to be able to use a closure similar to the one in the generated Wasmtime code.

Reply:
If someone can make it work I'd be happy with the change :shrug:

Reply:
Hah, no, I definitely wouldn't block this PR.